js/views/transactionsVw.js

```diff
@@ -281,6 +285,118 @@ module.exports = baseVw.extend({
     tabWrapper.append(orderShort.render().el);
   },
 
+  clickDownloadCSV: function(e){
+    var targBtn = $(e.target);
+    targBtn.closest('.btn').addClass('loading');
+    this.downloadCSV(targBtn);
+  },
+
+  downloadCSV: function(targBtn){
+    var self = this,
+        rawData = [],
+        calls = [],
+        dataCSV = '',
+        tempAnchor = document.createElement('a'),
+        dataBlob,
+        exportData = function(data){
+          "use strict";
+          dataCSV = Papa.unparse(data);
+          dataBlob = new Blob([dataCSV], {'type':'application\/octet-stream'});
+          tempAnchor.href = window.URL.createObjectURL(dataBlob);
+          tempAnchor.download = 'export.csv';
+          tempAnchor.click();
+        };
+
+    //clear existing data
+    this.currentExportData = [];
+
+    switch(this.state){
+      case "purchases":
+        rawData = this.purchasesCol.toJSON();
+        break;
+      case "sales":
+        rawData = this.salesCol.toJSON();
+        break;
+      case "cases":
+        rawData = this.casesCol.toJSON();
+        break;
+    }
+
+    $.each(rawData,(i, transaction)=> {
+      transaction.status = polyglot.t('transactions.OrderStatus'+transaction.status);
+      transaction.timestamp = new Date(transaction.timestamp * 1000);
+      transaction = __.omit(transaction, "thumbnail_hash", "btAve", "imageUrl", "order_date", "cCode", "displayPrice", "vendor", "transactionType");
+      calls.push(this.getTransactionData(transaction.order_id, transaction));
+    });
+
+    $.when.apply(null, calls)
```

Comment:
If any of the `calls` fail, then this `done()` handler will not be called and we will not follow through with the export. If that's indeed what we want (no partial export), then we need to at least do some cleanup in a fail handler. The fail handler will be called as soon as any call fails. At that point, we would want to abort all the other calls, since we're not doing anything with them, so it's pointless for them to go through. If there's enough of them, there may be some queue up that haven't even been sent yet.

```javascript
$.when.apply(null, calls)
        .fail(function(){
            calls.forEach((call) => {
                call.abort();
            });
        });
```